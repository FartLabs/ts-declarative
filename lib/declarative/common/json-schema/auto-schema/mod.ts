export * from "./auto-schema.ts";
