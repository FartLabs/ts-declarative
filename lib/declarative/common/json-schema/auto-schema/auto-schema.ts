import { Project } from "ts-morph";
import type { Class } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";
import { declarativeTypeInfo } from "#/lib/declarative/common/type-info/type-info.ts";
import type { JSONSchemaMask } from "#/lib/declarative/common/json-schema/json-schema.ts";
import { declarativeJSONSchema } from "#/lib/declarative/common/json-schema/json-schema.ts";

/**
 * createAutoSchemaDecoratorFactory creates a decorator factory that decorates
 * a value with a JSON Schema and automatically reads the type information
 * from the class.
 */
export async function createAutoSchemaDecoratorFactoryAt(
  { url }: ImportMeta,
): Promise<(mask?: JSONSchemaMask) => (target: Class) => Class> {
  const project = new Project({ useInMemoryFileSystem: true });
  const specifier = new URL(url);
  project.createSourceFile(
    specifier.toString(),
    await Deno.readTextFile(specifier),
  );

  return createDecoratorFactory({
    initialize: (mask?: JSONSchemaMask) => {
      return [
        declarativeTypeInfo(project, specifier),
        declarativeJSONSchema(mask),
      ];
    },
  });
}
