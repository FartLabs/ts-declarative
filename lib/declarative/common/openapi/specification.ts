import type { OpenAPIV3_1 } from "openapi-types";
import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";
import { jsonSchemaOf } from "#/lib/declarative/common/json-schema/json-schema.ts";
import { pathsObjectOf, reducePathsObject } from "./paths-object.ts";

/**
 * specificationOf returns the OpenAPI specification of the class.
 */
export function specificationOf<TClass extends Class>(
  target: TClass,
): OpenAPIV3_1.Document | undefined {
  return getPrototypeValue<ValueOpenAPISpecification>(target)?.specification;
}

/**
 * openapiSpec is the decorator for OpenAPI specification.
 */
export const openapiSpec: (
  options?: OpenAPISpecificationDecoratorOptions,
) => (target: Class) => Class = createOpenAPISpecificationDecoratorFactory();

/**
 * createOpenAPISpecificationDecoratorFactory is the factory function for the
 * OpenAPI decorator.
 */
export function createOpenAPISpecificationDecoratorFactory(): (
  options?: OpenAPISpecificationDecoratorOptions | undefined,
) => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (options?: OpenAPISpecificationDecoratorOptions) => {
      return [declarativeOpenAPISpecification(options)];
    },
  });
}

/**
 * OpenAPISpecificationDecoratorOptions is the options for the OpenAPI
 * specification decorator.
 */
export interface OpenAPISpecificationDecoratorOptions {
  /**
   * specification is the base OpenAPI specification.
   */
  specification?: Partial<OpenAPIV3_1.Document>;

  /**
   * resources are the resources of the OpenAPI specification.
   */
  resources?: Class[];
}

/**
 * defaultOpenAPI is the default OpenAPI version.
 */
export const defaultOpenAPI = "3.0.1";

/**
 * defaultOpenAPIInfoVersion is the default OpenAPI info version.
 */
export const defaultOpenAPIInfoVersion = "1.0.0";

/**
 * declarativeOpenAPISpecification is the declarative function for OpenAPI specification.
 */
export function declarativeOpenAPISpecification<
  TValue extends ValueOpenAPISpecification,
>(
  options?: OpenAPISpecificationDecoratorOptions,
): Declarative<TValue> {
  return (value, name) => {
    return {
      ...value,
      specification: {
        ...options?.specification,
        openapi: options?.specification?.openapi ?? defaultOpenAPI,
        info: {
          ...options?.specification?.info,
          title: options?.specification?.info?.title ?? name,
          version: options?.specification?.info?.version ??
            defaultOpenAPIInfoVersion,
        },
        paths: {
          ...options?.specification?.paths,
          ...(options?.resources
            ?.map((resource) => {
              const pathsObject = pathsObjectOf(resource);
              if (pathsObject === undefined) {
                throw new Error(`pathsObject is required for ${resource.name}`);
              }

              return pathsObject;
            })
            .reduce(reducePathsObject, {}) ?? {}),
        },
        components: {
          ...options?.specification?.components,
          schemas: {
            ...options?.specification?.components?.schemas,
            ...Object.fromEntries(
              options?.resources?.map((resource) => {
                const jsonSchema = jsonSchemaOf(resource);
                if (jsonSchema === undefined) {
                  throw new Error(
                    `jsonSchema is required for ${resource.name}`,
                  );
                }

                return [resource.name, jsonSchema];
              }) ?? [],
            ),
          },
        },
      },
    } as TValue;
  };
}

/**
 * ValueOpenAPISpecification is the value associated with an OpenAPI specification.
 */
export interface ValueOpenAPISpecification {
  /**
   * specification is the top-level object of the OpenAPI specification.
   */
  specification?: OpenAPIV3_1.Document;
}
