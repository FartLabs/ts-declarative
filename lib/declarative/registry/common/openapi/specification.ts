// import { slugify } from "@std/text/unstable-slugify";
import type { Class } from "#/lib/declarative/declarative.ts";
import { getPrototypeID } from "#/lib/declarative/declarative.ts";
import type { Registry } from "#/lib/declarative/registry/registry.ts";
import { jsonSchemaOf } from "#/lib/declarative/common/json-schema/json-schema.ts";

/**
 * OpenAPISpecification is a registry for OpenAPI resources that builds an
 * OpenAPI specification.
 */
export class OpenAPISpecification implements Registry<OpenAPIResourceOptions> {
  // deno-lint-ignore no-explicit-any
  public constructor(public specification: any) {}

  public register(target: Class, _options?: OpenAPIResourceOptions): void {
    const id = getPrototypeID(target);
    if (id === undefined) {
      throw new Error("Target must be a class");
    }

    const jsonSchema = jsonSchemaOf(target);
    if (jsonSchema === undefined) {
      throw new Error("Target must have a JSON schema");
    }

    // TODO: Define CRUD endpoints on base path.
    // TODO: Generate client library with oazapfts.
    // const path = options?.path ?? slugify(id);

    throw new Error("Not implemented.");
  }
}

/**
 * OpenAPIResourceOptions are options for registering an OpenAPI resource.
 */
export interface OpenAPIResourceOptions {
  path?: string;
}
