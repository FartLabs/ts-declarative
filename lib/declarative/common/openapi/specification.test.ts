import { assertEquals } from "@std/assert/equals";
import { assertSnapshot } from "@std/testing/snapshot";
import {
  createStandardMethodsDecoratorFactory,
  standardCreate,
  standardGet,
} from "#/lib/declarative/common/google-aip/methods/mod.ts";
import { createAutoSchemaDecoratorFactoryAt } from "#/lib/declarative/common/json-schema/auto-schema/auto-schema.ts";
import { openapiSpec, specificationOf } from "./specification.ts";

const autoSchema = await createAutoSchemaDecoratorFactoryAt(import.meta);

const kv = await Deno.openKv(":memory:");
const standardMethods = createStandardMethodsDecoratorFactory(kv);

@standardCreate({ kv })
@standardGet({ kv })
@autoSchema()
class Person {
  public constructor(public name: string) {}
}

@openapiSpec({ resources: [Person] })
class PersonAPI {}

Deno.test("openapiSpec decorator decorates value", () => {
  const specification = specificationOf(PersonAPI);
  assertEquals(specification?.info.title, "PersonAPI");
  assertEquals(specification?.paths, {
    "/people": {
      post: {
        description: "Creates Person",
        requestBody: {
          content: {
            "application/json": {
              schema: { $ref: "#/components/schemas/Person" },
            },
          },
          description: "The Person to create",
          required: true,
        },
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Person" },
              },
            },
            description: "The created Person",
          },
        },
      },
    },
    "/people/{person}": {
      get: {
        description: "Gets Person",
        parameters: [
          {
            in: "path",
            name: "person",
            required: true,
            schema: { type: "string" },
          },
        ],
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Person" },
              },
            },
            description: "Got Person",
          },
        },
      },
    },
  });
});

@standardMethods()
@autoSchema()
class Cat {
  public constructor(public name: string) {}
}

@standardMethods()
@autoSchema()
class Dog {
  public constructor(public name: string) {}
}

@openapiSpec({ resources: [Cat, Dog] })
class ZooAPI {}

Deno.test("openapiSpec decorator with multiple resources", async (t) => {
  const specification = specificationOf(ZooAPI);
  await assertSnapshot(t, specification);
});
