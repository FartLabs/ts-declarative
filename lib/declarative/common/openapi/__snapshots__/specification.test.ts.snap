export const snapshot = {};

snapshot[`openapiSpec decorator with multiple resources 1`] = `
{
  components: {
    schemas: {
      Cat: {
        [Symbol(TypeBox.Kind)]: "Object",
        properties: {
          name: {
            [Symbol(TypeBox.Kind)]: "String",
            type: "string",
          },
        },
        required: [
          "name",
        ],
        type: "object",
      },
      Dog: {
        [Symbol(TypeBox.Kind)]: "Object",
        properties: {
          name: {
            [Symbol(TypeBox.Kind)]: "String",
            type: "string",
          },
        },
        required: [
          "name",
        ],
        type: "object",
      },
    },
  },
  info: {
    title: "ZooAPI",
    version: "1.0.0",
  },
  openapi: "3.0.1",
  paths: {
    "/cats": {
      get: {
        description: "Lists Cats",
        parameters: [
          {
            in: "query",
            name: "page_size",
          },
          {
            in: "query",
            name: "page_token",
          },
        ],
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  items: {
                    "\$ref": "#/components/schemas/Cat",
                  },
                  type: "array",
                },
              },
            },
            description: "List of Cats",
          },
        },
      },
      post: {
        description: "Creates Cat",
        requestBody: {
          content: {
            "application/json": {
              schema: {
                "\$ref": "#/components/schemas/Cat",
              },
            },
          },
          description: "The Cat to create",
          required: true,
        },
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  "\$ref": "#/components/schemas/Cat",
                },
              },
            },
            description: "The created Cat",
          },
        },
      },
    },
    "/cats/{cat}": {
      delete: {
        description: "Deletes Cat",
        parameters: [
          {
            in: "path",
            name: "cat",
            required: true,
          },
        ],
        responses: {
          "200": {
            description: "The deleted Cat",
          },
        },
      },
      get: {
        description: "Gets Cat",
        parameters: [
          {
            in: "path",
            name: "cat",
            required: true,
            schema: {
              type: "string",
            },
          },
        ],
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  "\$ref": "#/components/schemas/Cat",
                },
              },
            },
            description: "Got Cat",
          },
        },
      },
      post: {
        description: "Updates Cat",
        parameters: [
          {
            in: "path",
            name: "cat",
            required: true,
          },
        ],
        requestBody: {
          content: {
            "application/json": {
              schema: {
                "\$ref": "#/components/schemas/Cat",
              },
            },
          },
          description: "The Cat to update",
          required: true,
        },
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  "\$ref": "#/components/schemas/Cat",
                },
              },
            },
            description: "The updated Cat",
          },
        },
      },
    },
    "/dogs": {
      get: {
        description: "Lists Dogs",
        parameters: [
          {
            in: "query",
            name: "page_size",
          },
          {
            in: "query",
            name: "page_token",
          },
        ],
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  items: {
                    "\$ref": "#/components/schemas/Dog",
                  },
                  type: "array",
                },
              },
            },
            description: "List of Dogs",
          },
        },
      },
      post: {
        description: "Creates Dog",
        requestBody: {
          content: {
            "application/json": {
              schema: {
                "\$ref": "#/components/schemas/Dog",
              },
            },
          },
          description: "The Dog to create",
          required: true,
        },
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  "\$ref": "#/components/schemas/Dog",
                },
              },
            },
            description: "The created Dog",
          },
        },
      },
    },
    "/dogs/{dog}": {
      delete: {
        description: "Deletes Dog",
        parameters: [
          {
            in: "path",
            name: "dog",
            required: true,
          },
        ],
        responses: {
          "200": {
            description: "The deleted Dog",
          },
        },
      },
      get: {
        description: "Gets Dog",
        parameters: [
          {
            in: "path",
            name: "dog",
            required: true,
            schema: {
              type: "string",
            },
          },
        ],
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  "\$ref": "#/components/schemas/Dog",
                },
              },
            },
            description: "Got Dog",
          },
        },
      },
      post: {
        description: "Updates Dog",
        parameters: [
          {
            in: "path",
            name: "dog",
            required: true,
          },
        ],
        requestBody: {
          content: {
            "application/json": {
              schema: {
                "\$ref": "#/components/schemas/Dog",
              },
            },
          },
          description: "The Dog to update",
          required: true,
        },
        responses: {
          "200": {
            content: {
              "application/json": {
                schema: {
                  "\$ref": "#/components/schemas/Dog",
                },
              },
            },
            description: "The updated Dog",
          },
        },
      },
    },
  },
}
`;
